action/envelope.go

```diff
@@ -230,10 +239,100 @@ func (elp *Envelope) Serialize() []byte {
 	return byteutil.Must(proto.Marshal(elp.Proto()))
 }
 
-// Hash returns the hash value of SealedEnvelope.
+// Hash returns the hash value of Envelope.
 func (elp *Envelope) Hash() hash.Hash256 {
+	if elp.IsExternal() {
+		return elp.rawHash
+	}
 	return hash.Hash256b(elp.Serialize())
 }
 
 // SetNonce sets the nonce value
 func (elp *Envelope) SetNonce(n uint64) { elp.nonce = n }
+
+// InvalidChainID detects invalid chain ID
+func (elp *Envelope) InvalidChainID() bool {
+	return elp.chainID != 0 && elp.chainID != config.ExternChainID()
+}
+
+// IsExternal returns whether the action is external or not
+func (elp *Envelope) IsExternal() bool {
+	return elp.chainID == config.ExternChainID()
+}
+
+// ValidateExternal validates an external action
+func (elp *Envelope) ValidateExternal(sig []byte) error {
+	var (
+		tx  rlpTransaction
+		err error
+	)
+	switch act := elp.payload.(type) {
+	case *Transfer:
+		tx = (*Transfer)(act)
+	case *Execution:
+		tx = (*Execution)(act)
+	default:
+		return errors.Errorf("invalid action type %T not supported", act)
+	}
+
+	elp.rawHash, elp.hash, err = computeRlpHash(tx, elp.chainID, sig)
+	return err
+}
+
+type rlpTransaction interface {
+	Nonce() uint64
+	GasPrice() *big.Int
+	GasLimit() uint64
+	Recipient() string
+	Amount() *big.Int
+	Payload() []byte
+}
+
+func computeRlpHash(act rlpTransaction, chainID uint32, sig []byte) (hash.Hash256, hash.Hash256, error) {
```

Comment:
(hash.Hash256, hash.Hash256, error)
better to assign varriable name

Reply:
ok